src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,421 @@
+        /// </summary>
+        /// <param name="serialPortStream">A stream to an Arduino/Firmata device</param>
+        public ArduinoBoard(Stream serialPortStream)
+        {
+            _dataStream = serialPortStream;
+            _spiEnabled = 0;
+            _supportedPinConfigurations = new List<SupportedPinConfiguration>();
+            _firmwareName = string.Empty;
+        }
+
+        /// <summary>
+        /// Creates an instance of the Arduino board connection connected to a serial port
+        /// </summary>
+        /// <param name="portName">Port name. On Windows, this is usually "COM3" or "COM4" for an Arduino attached via USB.
+        /// On Linux, possible values include "/dev/ttyAMA0", "/dev/serial0", "/dev/ttyUSB1", etc.</param>
+        /// <param name="baudRate">Baudrate to use. It is recommended to use at least 115200 Baud.</param>
+        public ArduinoBoard(string portName, int baudRate)
+        {
+            _serialPort = new SerialPort(portName, baudRate);
+            _serialPort.Open();
+            _dataStream = _serialPort.BaseStream;
+            _supportedPinConfigurations = new List<SupportedPinConfiguration>();
+            _firmwareName = string.Empty;
+        }
+
+        /// <summary>
+        /// Default pin numbering scheme of this board. Always <see cref="PinNumberingScheme.Logical"/>, because the Arduino has only one pin numbering scheme.
+        /// </summary>
+        public PinNumberingScheme DefaultPinNumberingScheme
+        {
+            get
+            {
+                // We have only one scheme
+                return PinNumberingScheme.Logical;
+            }
+        }
+
+        /// <summary>
+        /// Attach to this event to retrieve log messages
+        /// </summary>
+        public event Action<string, Exception?>? LogMessages;
+
+        /// <summary>
+        /// Searches the given list of com ports for a firmata device.
+        /// </summary>
+        /// <param name="comPorts">List of com ports. See <see cref="GetSerialPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        public static ArduinoBoard? FindBoard(List<string> comPorts, List<int> baudRates)
+        {
+            foreach (var port in comPorts)
+            {
+                foreach (var baud in baudRates)
+                {
+                    ArduinoBoard? b = null;
+                    try
+                    {
+                        b = new ArduinoBoard(port, baud);
+                        b.Initialize();
+                        return b;
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
```

Comment:
should you dispose on any exception?

Reply:
There shouldn't be any other exceptions than these. If there are, it's most likely a programming error (i.e. a NullReferenceException) when I really don't care.